usb/usb_host_hid/hid_host.c

```diff
@@ -0,0 +1,1474 @@
+        ESP_ERROR_CHECK( hid_host_interface_list_create(hid_device, config_desc) );
+    } else {
+        usb_host_device_close(s_hid_driver->client_handle, dev_hdl);
+        ESP_LOGW(TAG, "No HID device at USB port %d", dev_addr);
+    }
+
+    return is_hid_device;
+}
+
+/**
+ * @brief USB device was removed we need to shutdown HID Interface
+ *
+ * @param[in] hid_dev_handle    Handle of the HID devive to close
+ * @param[in] force             To shutdown HID interface immediately
+ * @return esp_err_t
+ */
+static esp_err_t hid_host_interface_shutdown(hid_host_device_handle_t hid_dev_handle,
+        bool force)
+{
+    hid_iface_t *iface = get_iface_by_handle(hid_dev_handle);
+
+    HID_RETURN_ON_INVALID_ARG(iface);
+
+    iface->state = HID_INTERFACE_STATE_WAIT_USER_DELETION;
+
+    // Not force shutdown and Interface was started by user
+    if (!force && iface->user_cb) {
+        hid_host_user_interface_callback(iface, HID_HOST_INTERFACE_EVENT_DISCONNECTED);
+        return ESP_OK;
+    }
+
+    // Interface has not been opened, delete it right away
+    return hid_host_device_close(iface);
+}
+
+/**
+ * @brief Destroy a list of available interfaces in RAM
+ *
+ * @return esp_err_t
+ */
+static esp_err_t hid_host_interface_list_destroy(void)
+{
+    bool active_shutdown = false;
+
+    hid_iface_t *iface = STAILQ_FIRST(&hid_ifaces_tailq);
+    hid_iface_t *tmp = NULL;
+
+    while (iface != NULL) {
+        tmp = STAILQ_NEXT(iface, tailq_entry);
+        // free interface
+        if (HID_INTERFACE_STATE_NOT_INITIALIZED != iface->state) {
+            active_shutdown = true;
+            ESP_ERROR_CHECK( hid_host_device_close(iface) );
+            ESP_ERROR_CHECK( hid_host_interface_shutdown(iface, true));
+        }
+
+        HID_ENTER_CRITICAL();
+        iface->state = HID_INTERFACE_STATE_NOT_INITIALIZED;
+        STAILQ_REMOVE(&hid_ifaces_tailq, iface, hid_interface, tailq_entry);
+        free(iface);
+        HID_EXIT_CRITICAL();
+
+        iface = tmp;
+    }
+
+    if (active_shutdown) {
+        ESP_LOGE(TAG, "Shutdown active HID device.");
+    }
+
+    return ESP_OK;
+}
+
+/**
+ * @brief Deinit USB device by handle
+ *
+ * @param[in] dev_hdl   USB device handle
+ * @return uint8_t USB device address
+ */
+static uint8_t hid_host_device_disconnected(usb_device_handle_t dev_hdl)
+{
+    uint8_t dev_addr = 0;
+    hid_device_t *hid_device = get_hid_device_by_handle(dev_hdl);
+    // Device should be in the list
+    assert(hid_device);
+    // Keep USB address to return it
+    dev_addr = hid_device->dev_addr;
+
+    hid_iface_t *iface = STAILQ_FIRST(&hid_ifaces_tailq);
+    hid_iface_t *tmp = NULL;
+
+    while (iface != NULL) {
+        tmp = STAILQ_NEXT(iface, tailq_entry);
+        if (iface->parent && (iface->parent->dev_addr == dev_addr)) {
+            ESP_ERROR_CHECK( hid_host_device_close(iface) );
+            ESP_ERROR_CHECK( hid_host_interface_shutdown(iface, false) );
```

Comment:
Are we sure we want to free the interface and device automatically? As soon as we the interface object is freed, the `hid_host_device_handle_t` in the user's code is no longer valid (will point to an already freed object).

The safer way to handle this is to keep the interface object allocated (but have a state variable in the interface indicating that the interface is gone). The interface object remains allocated until the user explicitly calls `hid_host_device_close()`). Doing this will also save us from having to call `get_iface_by_handle()` at the start of every HID interface function, as we can assume that a provided handle will always have an allocated object behind it.

Reply:
Yeah....that the point :(
Now, there are two callback are available:
1. device connected (for every HID interface)
2. device event: input report, transfer error, device gone

If the user doesn't open a device after receiving device connected event, the callback with 'device gone' will not reach user. So that means, that there are two possible way of closing the device (interface in this case): 
1. If if was opened by user, close everything we can inside the driver, send event to user for delete object.
2. If it was not opened, then close everything we can inside the driver and delete object right now.

That is exactly the way as it implemented right now. 

But from some point, I am concerned about two callbacks and I want to make it simpler and merge them to one in the future. 
That would be a breaking change then, but I don't see the benefits from two callback right now (at least, anymore after several process of refactoring and eliminating other logic).